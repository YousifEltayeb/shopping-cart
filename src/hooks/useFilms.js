import { useEffect, useState } from "react";
const API_KEY = import.meta.env.VITE_TMBD_KEY;
const useFilms = () => {
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(true);
  const [fetchedProducts, setFetchedProducts] = useState([]);
  useEffect(() => {
    (async () => {
      // if we already called the api and data is in products no need to call again
      if (fetchedProducts.length > 0) {
        setLoading(false);
        return;
      }
      const url =
        "https://api.themoviedb.org/3/discover/movie?with_genres=10751,16&sort_by=vote_average.desc&language=en-US&page=1&vote_count.gte=200";
      const options = {
        method: "GET",
        headers: {
          accept: "application/json",
          Authorization: `Bearer ${API_KEY}`,
        },
      };
      // fetch films
      fetch(url, options)
        .then((response) => {
          if (response.status >= 400) {
            throw new Error("server error");
          }
          return response.json();
        })
        .then((json) => {
          const newProductsArray = [];
          const fetchedFilms = json.results;
          fetchedFilms.forEach((film) => {
            const imgURL =
              "https://image.tmdb.org/t/p/w780" + film.poster_path || "";
            const newProduct = {
              id: film.id,
              title: film.title,
              posterPath: imgURL,
              inCart: false,
              price: 20,
              quantity: 0,
            };
            newProductsArray.push(newProduct);
          });
          setFetchedProducts(newProductsArray);
        })
        .catch((err) => setError(err))
        .finally(() => setLoading(false));
    })();
  }, [fetchedProducts.length, setFetchedProducts]);
  return { fetchedProducts, error, loading };
};

export default useFilms;
