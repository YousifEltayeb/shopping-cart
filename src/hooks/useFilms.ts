import { useEffect, useState } from "react";
const API_KEY = import.meta.env.VITE_TMBD_KEY;
interface Product {
  id: number;
  title: string;
  posterPath: string;
  inCart: false;
  price: number;
  quantity: number;
}
const useFilms = () => {
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(true);
  const [fetchedProducts, setFetchedProducts] = useState<Product[]>([]);
  useEffect(() => {
    (async () => {
      // if we already called the api and data is in products no need to call again
      if (fetchedProducts.length > 0) {
        setLoading(false);
        return;
      }
      const url =
        "https://api.themoviedb.org/3/discover/movie?with_genres=10751,16&sort_by=vote_average.desc&language=en-US&page=1&vote_count.gte=200";
      const options = {
        method: "GET",
        headers: {
          accept: "application/json",
          Authorization: `Bearer ${API_KEY}`,
        },
      };
      // fetch films
      fetch(url, options)
        .then((response) => {
          if (response.status >= 400) {
            throw new Error("server error");
          }
          return response.json();
        })
        .then((json) => {
          const newProductsArray: Product[] = [];
          const fetchedFilms = json.results;
          fetchedFilms.forEach(
            (film: { id: number; title: string; poster_path: string }) => {
              const imgURL =
                "https://image.tmdb.org/t/p/w780" + film.poster_path || "";
              const newProduct: Product = {
                id: film.id,
                title: film.title,
                posterPath: imgURL,
                inCart: false,
                price: 20,
                quantity: 0,
              };
              newProductsArray.push(newProduct);
            },
          );
          setFetchedProducts(newProductsArray);
        })
        .catch((err) => setError(err))
        .finally(() => setLoading(false));
    })();
  }, [fetchedProducts.length, setFetchedProducts]);
  return { fetchedProducts, error, loading };
};

export default useFilms;
