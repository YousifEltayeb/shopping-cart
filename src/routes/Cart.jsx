import styled from "styled-components";
import { useOutletContext, Link } from "react-router";
const Main = styled.main`
  display: grid;
  grid-template-columns: 1fr 1fr;
`;
const Items = styled.section`
  display: grid;
  grid-template-rows: repeat(auto-fill, 12rem);
  padding: 18px;
  margin-left: 50px;
  gap: 18px;
`;
const Item = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-template-rows: 12rem;
  align-items: center;
`;
const Poster = styled.img`
  width: 100%;
  height: 100%;
  object-fit: scale-down;
`;
const Checkout = styled.button`
  background-color: red;
  width: 18rem;
`;
const OrderDetails = styled.div`
  display: flex;
  gap: 18px;
`;
const Order = styled.section`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 64px;
`;
const EmptyCart = styled.main``;
const Cart = () => {
  const [products] = useOutletContext();
  const inCartProducts = products.filter((product) => product.inCart === true);
  const inCartQuantity = inCartProducts.reduce(
    (currentTotal, { quantity }) => currentTotal + quantity,
    0,
  );
  const orderTotal = inCartQuantity * 20;
  if (inCartProducts.length === 0) {
    return (
      <EmptyCart data-testid="empty-cart" aria-label="Cart">
        <h1>Cart is empty</h1>
        <h2>
          <Link to="/store"> return to store to addd products</Link>
        </h2>
      </EmptyCart>
    );
  }
  return (
    <Main data-testid="non-empty-cart" aria-label="Cart">
      <Items>
        {inCartProducts.map((product) => {
          return (
            <Item key={product.id}>
              <Poster src={product.posterPath} alt="film poster" width="auto" />
              <h4>{product.title}</h4>
              <div>{product.quantity}x</div>
              <div>
                <b>{product.price}$</b>
              </div>
            </Item>
          );
        })}
      </Items>
      <Order>
        <h2>Order Summery</h2>
        <OrderDetails>
          <h3>Total ({inCartQuantity} items)</h3>
          <h3 data-testid="total">{orderTotal}$</h3>
        </OrderDetails>
        <Checkout>Checkout</Checkout>
      </Order>
    </Main>
  );
};

export default Cart;
