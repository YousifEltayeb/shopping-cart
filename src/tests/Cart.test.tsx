import { beforeEach, describe, it, expect, vi, Mock } from "vitest";
import { render, screen } from "@testing-library/react";
import { useProducts } from "../hooks/useFilms";
import { MemoryRouter } from "react-router";
import Cart from "../routes/Cart";
vi.mock("../hooks/useFilms", () => ({
  useProducts: vi.fn(),
}));
describe("Cart tests", () => {
  const mockedUseProducts = useProducts as Mock;
  beforeEach(() => {
    vi.clearAllMocks(); // Reset all mocked calls between tests
  });
  it("Empty cart displays empty-cart element", async () => {
    const products = [
      {
        inCart: false,
      },
    ];

    mockedUseProducts.mockReturnValue({
      products,
    });
    const { container } = render(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });
  it("Display products in cart", async () => {
    const products = [
      {
        id: 0,
        title: "Sprited Away",
        posterPath: "sprit.jpg",
        inCart: true,
        price: 20,
        quantity: 1,
      },
      {
        id: 1,
        title: "The Wild Robot",
        posterPath: "robot.jpg",
        inCart: true,
        price: 20,
        quantity: 2,
      },
      {
        id: 2,
        title: "Lion King",
        posterPath: "lion.jpg",
        inCart: true,
        price: 20,
        quantity: 1,
      },
    ];
    mockedUseProducts.mockReturnValue({ products });
    render(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>,
    );
    // total should be 80
    expect(screen.getByTestId("total").textContent).toBe("80$");
    // film titles should be in the document
    expect(screen.getByText("Lion King")).toBeInTheDocument();
    expect(screen.getByText("The Wild Robot")).toBeInTheDocument();
    expect(screen.getByText("Sprited Away")).toBeInTheDocument();
    // posters should be displayed
    expect(screen.getByAltText("Lion King poster")).toBeInTheDocument();
    expect(screen.getByAltText("The Wild Robot poster")).toBeInTheDocument();
    expect(screen.getByAltText("Sprited Away poster")).toBeInTheDocument();
  });
});
