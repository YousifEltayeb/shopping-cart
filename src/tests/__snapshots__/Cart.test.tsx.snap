// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart tests > Empty cart displays empty-cart element 1`] = `
<div>
  <main
    aria-label="Cart"
    class="sc-kfeOyU"
    data-testid="empty-cart"
  >
    <h1>
      Cart is empty
    </h1>
    <h2>
      <a
        data-discover="true"
        href="/store"
      >
         return to store to addd products
      </a>
    </h2>
  </main>
</div>
`;
